src/java/edu/slu/action/ObjectAction.java

```diff
@@ -527,116 +527,167 @@ public void batchSaveFromCopy() throws UnsupportedEncodingException, IOException
+
+    private JSONArray getAllDescendants(List<DBObject> ls_versions, JSONObject keyObj, JSONArray discoveredDescendants){
+        JSONArray nextIDarr = new JSONArray();
+        if(keyObj.getJSONObject("__rerum").getJSONObject("history").getJSONArray("next").size() > 0){
+            //essentially, do nothing.  This branch is done.
+        }
+        else{
+            //The provided object has nexts, get them to add them to known descendants then check their descendants.
+            nextIDarr = keyObj.getJSONObject("__rerum").getJSONObject("history").getJSONArray("next");
         }
+        //@cubap what should I do if I find malformed objects here?  Do we need a try-catch?
+        for(int m=0; m<nextIDarr.size(); m++){ //For each id in the array
+            String nextID = nextIDarr.getString(m);
+            for (DBObject thisVersion : ls_versions) {
+                JSONObject thisObject = JSONObject.fromObject(thisVersion);
+                if(thisObject.getString("@id").equals(nextID)){ //If it is equal, add it to the known descendants
+                    discoveredDescendants.add(thisObject);
+                    //Recurse with what you have discovered so far and this object as the new keyObj
+                    getAllDescendants(ls_versions, thisObject, discoveredDescendants);
+                    break;
+                }
+            }
+        }
+        return discoveredDescendants;
     }
     
     /**
-     * Loads all derivative versions from the `prime` object. Used for filtering
-     * in other methods. May be replaced later with more optimized logic.
-     * @param  http_request Servlet request for relatives
+     * Internal private method to loads all derivative versions from the `root` object. It should always receive a reliable object, not one from the user.
+     * Used to resolve the history tree for storing into memory.
+     * @param  obj A JSONObject to find all versions of.  If it is root, make sure to prepend it to the result.  If it isn't root, query for root from the ID
+     * found in prime using that result as a reliable root object. 
      * @return All versions from the store of the object in the request
      * @throws Exception 
      */
-    private List<DBObject> getAllVersions(HttpServletRequest http_request) throws Exception {
+    private List<DBObject> getAllVersions(JSONObject obj) throws Exception {
         List<DBObject> ls_versions = null;
-        if(!methodApproval(request, "get")){
-            // TODO: include link to API documentation in error response
-            writeErrorResponse("Unable to retrieve objects; wrong method type.", HttpServletResponse.SC_BAD_REQUEST);
-            return ls_versions;
+        BasicDBObject rootObj;
+        BasicDBObject query = new BasicDBObject();
+        BasicDBObject queryForRoot = new BasicDBObject();  
+        String primeID;
+        //@cubap what should I do if I find malformed objects here?  Do we need a try-catch?
```

Comment:
I'm not even sure how you would recognize a malformed object here, but I would treat it like above.

Reply:
private actions will presumably not be fed malformed objects.  If the object is malformed, it will be detected before it ever reaches these functions.  I will remove these superfluous comments.